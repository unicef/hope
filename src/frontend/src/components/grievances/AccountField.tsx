import { Grid2 as Grid, IconButton, Button } from '@mui/material';
import { Delete, Add } from '@mui/icons-material';
import { useLocation } from 'react-router-dom';
import { Field, FieldArray } from 'formik';
import { useTranslation } from 'react-i18next';
import { FormikTextField } from '@shared/Formik/FormikTextField';
import { LabelizedField } from '@core/LabelizedField';
import { getIndexForId } from './utils/helpers';
import React, { Fragment, ReactElement } from 'react';
import { FormikSelectField } from '@shared/Formik/FormikSelectField';
import { Account } from '@restgenerated/models/Account';


export interface AccountProps {
  id: string;
  baseName: string;
  onDelete;
  isEdited?: boolean;
  account?: Account;
  values;
  accountTypeChoices: Record<string, any>[];
  accountFinancialInstitutionChoices: Record<string, any>[];
}

export function AccountField({
                               id,
                               baseName,
                               onDelete,
                               isEdited,
                               account,
                               values,
                               accountTypeChoices,
                               accountFinancialInstitutionChoices,
                             }: AccountProps): ReactElement {
  const { t } = useTranslation();

  const accountFieldName = `${baseName}.${getIndexForId(
    values[baseName],
    id,
  )}`;

  const location = useLocation();
  const isEditTicket = location.pathname.indexOf('edit-ticket') !== -1;
  const dataFields = account?.dataFields || {};

  const dynamicFieldsName = `${accountFieldName}.dynamicFields`;

  return (
    <>
      <Grid size={{ xs: 11 }}/>
      {!isEdited ? (
        <Grid size={{ xs: 1 }}>
          <IconButton disabled={isEditTicket} onClick={onDelete}>
            <Delete/>
          </IconButton>
        </Grid>
      ) : null}

      <Fragment key="type">
        <Grid size={{ xs: 4 }}>
          <LabelizedField
            label={t('Account Item')}
            value="type"
          />
        </Grid>

        {account ? (
          <>
            <Grid size={{ xs: 4 }}>
              <LabelizedField
                label={t('Current Value')}
                value={account?.accountType || ''}
              />
            </Grid>
            <Grid size={{ xs: 3 }}>
              <LabelizedField
                label={t('New Value')}
                value={account?.accountType || ''}
              />
            </Grid>
          </>
        ) : (
          <Grid size={{ xs: 8 }}>
            <Field
              name={`${accountFieldName}.name`}
              variant="outlined"
              label={t('New Value')}
              component={FormikSelectField}
              choices={accountTypeChoices}
              required
            />
          </Grid>
        )}
      </Fragment>

      {!account ? (
        <>
          <Fragment key="number">
            <Grid size={{ xs: 4 }}>
              <LabelizedField
                label={t('Account Item')}
                value="number"
              />
            </Grid>
            <Grid size={{ xs: 8 }}>
              <Field
                fullWidth
                name={`${accountFieldName}.number`}
                variant="outlined"
                label={t('New Value')}
                component={FormikTextField}
                required
              />
            </Grid>
          </Fragment>
          <Fragment key="financial_institution">
            <Grid size={{ xs: 4 }}>
              <LabelizedField
                label={t('Account Item')}
                value="financial_institution"
              />
            </Grid>
            <Grid size={{ xs: 8 }}>
              <Field
                name={`${accountFieldName}.financial_institution`}
                variant="outlined"
                label={t('New Value')}
                component={FormikSelectField}
                choices={accountFinancialInstitutionChoices}
                required
              />
            </Grid>
          </Fragment>
        </>
      ) : (
        <>
          {Object.entries(dataFields).map(([key, value]) => {
            let displayValue = String(value);
              const isFinancialInstitutionField = key === 'financial_institution';
              if (
                isFinancialInstitutionField &&
                Array.isArray(accountFinancialInstitutionChoices)
              ) {
                const choice = accountFinancialInstitutionChoices.find(
                  (c: any) => c.value === value,
                );
                displayValue = choice ? choice.name : String(value);
              }
              const fieldProps = {
                component: isFinancialInstitutionField ? FormikSelectField : FormikTextField,
                ...(isFinancialInstitutionField ? { choices: accountFinancialInstitutionChoices } : {}),
              };

            return (
            <Fragment key={key}>
              <Grid size={{ xs: 4 }}>
                <LabelizedField
                  label={t('Account Item')}
                  value={String(key)}
                />
              </Grid>
              <Grid size={{ xs: 4 }}>
                <LabelizedField
                  label={t('Current Value')}
                  value={displayValue}
                />
              </Grid>
              <Grid size={{ xs: 3 }}>
                  <Field
                      name={`${accountFieldName}.${key}`}
                      fullWidth
                      variant="outlined"
                      label={t('New Value')}
                      required={!isEditTicket}
                      disabled={isEditTicket}
                      {...fieldProps}
                    />
              </Grid>
            </Fragment>
          );
})}
        </>
      )
      }
      {/* --- Dynamic Fields Section --- */}
      <FieldArray name={dynamicFieldsName}>
        {/* eslint-disable-next-line @typescript-eslint/unbound-method*/}
        {({ push, remove, form }) => (
          <>
            {form.values[baseName][getIndexForId(values[baseName], id)]?.dynamicFields?.map(
              (field, idx) => (
                <Fragment key={idx}>
                  <Grid size={{ xs: 4 }}>
                    <Field
                      name={`${dynamicFieldsName}.${idx}.key`}
                      component={FormikTextField}
                      label={t('Field Name')}
                      variant="outlined"
                      fullWidth
                      required
                    />
                  </Grid>
                  <Grid size={{ xs: 7 }}>
                    <Field
                      name={`${dynamicFieldsName}.${idx}.value`}
                      component={FormikTextField}
                      label={t('Field Value')}
                      variant="outlined"
                      fullWidth
                      required
                    />
                  </Grid>
                  <Grid size={{ xs: 1 }}>
                    <IconButton onClick={() => remove(idx)}>
                      <Delete/>
                    </IconButton>
                  </Grid>
                </Fragment>
              ),
            )}
            <Grid size={{ xs: 12 }}>
              <Button
                variant="outlined"
                startIcon={<Add/>}
                onClick={() => push({ key: '', value: '' })}
              >
                {t('Add Field')}
              </Button>
            </Grid>
          </>
        )}
      </FieldArray>
      {/* --- End Dynamic Fields --- */}
    </>
  );
}