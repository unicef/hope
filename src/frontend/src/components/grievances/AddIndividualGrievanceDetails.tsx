import { Box, Button, Grid, Typography } from '@mui/material';
import * as React from 'react';
import capitalize from 'lodash/capitalize';
import { useTranslation } from 'react-i18next';
import { useSnackbar } from '@hooks/useSnackBar';
import { GRIEVANCE_TICKET_STATES } from '@utils/constants';
import { getFlexFieldTextValue, renderBoolean } from '@utils/utils';
import {
  GrievanceTicketDocument,
  GrievanceTicketQuery,
  useAllAddIndividualFieldsQuery,
  useApproveAddIndividualDataChangeMutation,
} from '@generated/graphql';
import { useConfirmation } from '@core/ConfirmationDialog';
import { LabelizedField } from '@core/LabelizedField';
import { LoadingComponent } from '@core/LoadingComponent';
import { Title } from '@core/Title';
import { ApproveBox } from './GrievancesApproveSection/ApproveSectionStyles';

export function AddIndividualGrievanceDetails({
  ticket,
  canApproveDataChange,
}: {
  ticket: GrievanceTicketQuery['grievanceTicket'];
  canApproveDataChange: boolean;
}): React.ReactElement {
  const { t } = useTranslation();
  const { data, loading } = useAllAddIndividualFieldsQuery();
  const [mutate] = useApproveAddIndividualDataChangeMutation();
  const confirm = useConfirmation();
  const { showMessage } = useSnackbar();
  if (loading) {
    return <LoadingComponent />;
  }
  if (!data) {
    return null;
  }
  const fieldsDict = data.allAddIndividualsFieldsAttributes.reduce(
    (previousValue, currentValue) => ({
      ...previousValue,
      [currentValue?.name]: currentValue,
    }),
    {},
  );

  const individualData = {
    ...ticket.addIndividualTicketDetails?.individualData,
  };
  const documents = individualData?.documents;
  const identities = individualData?.identities;
  delete individualData.documents;
  delete individualData.identities;
  const flexFields = individualData?.flex_fields;
  delete individualData?.flex_fields;
  delete individualData.documents;
  delete individualData.identities;
  const labels =
    Object.entries(individualData || {}).map(([key, value]) => {
      let textValue = value;

      const fieldAttribute = fieldsDict[key];
      if (fieldAttribute.type === 'BOOL') {
        textValue = renderBoolean(value as boolean);
      }
      if (fieldAttribute.type === 'SELECT_ONE') {
        textValue =
          fieldAttribute.choices.find((item) => item.value === value)
            ?.labelEn || '-';
      }
      if (Array.isArray(value)) {
        textValue = value.map((el) => capitalize(el)).join(', ');
      }
      return (
        <Grid key={key} item xs={6}>
          <LabelizedField
            label={key === 'sex' ? t('GENDER') : key.replace(/_/g, ' ')}
            value={<span>{textValue as React.ReactNode}</span>}
          />
        </Grid>
      );
    }) || [];

  const flexFieldLabels =
    Object.entries(flexFields || {}).map(
      ([key, value]: [string, string | string[]]) => (
        <Grid key={key} item xs={6}>
          <LabelizedField
            label={key.replaceAll('_i_f', '').replace(/_/g, ' ')}
            value={getFlexFieldTextValue(key, value, fieldsDict[key])}
          />
        </Grid>
      ),
    ) || [];
  const documentLabels =
    documents?.map((item) => (
      <Grid key={item?.country + item?.key} item xs={6}>
        <LabelizedField
          label={item?.key?.replace(/_/g, ' ')}
          value={item.number}
        />
      </Grid>
    )) || [];
  const identityLabels =
    identities?.map((item) => {
      const partner = item.partner || item.agency; // For backward compatibility
      return (
        <Grid key={item.country + partner} item xs={6}>
          <LabelizedField label={partner} value={item.number} />
        </Grid>
      );
    }) || [];
  const allLabels = [
    ...labels,
    ...flexFieldLabels,
    ...documentLabels,
    ...identityLabels,
  ];

  let dialogText = t(
    'You did not approve the following add individual data. Are you sure you want to continue?',
  );
  if (!ticket.addIndividualTicketDetails.approveStatus) {
    dialogText = t(
      'You are approving the following Add individual data. Are you sure you want to continue?',
    );
  }

  return (
    <ApproveBox>
      <Title>
        <Box display="flex" justifyContent="space-between">
          <Typography variant="h6">{t('Individual Data')}</Typography>
          {canApproveDataChange && (
            <Button
              data-cy="button-approve"
              onClick={() =>
                confirm({
                  title: t('Warning'),
                  content: dialogText,
                }).then(async () => {
                  try {
                    await mutate({
                      variables: {
                        grievanceTicketId: ticket.id,
                        approveStatus:
                          !ticket.addIndividualTicketDetails.approveStatus,
                      },
                      refetchQueries: () => [
                        {
                          query: GrievanceTicketDocument,
                          variables: { id: ticket.id },
                        },
                      ],
                    });
                    if (ticket.addIndividualTicketDetails.approveStatus) {
                      showMessage(t('Changes Disapproved'));
                    }
                    if (!ticket.addIndividualTicketDetails.approveStatus) {
                      showMessage(t('Changes Approved'));
                    }
                  } catch (e) {
                    e.graphQLErrors.map((x) => showMessage(x.message));
                  }
                })
              }
              variant={
                ticket.addIndividualTicketDetails?.approveStatus
                  ? 'outlined'
                  : 'contained'
              }
              color="primary"
              disabled={ticket.status !== GRIEVANCE_TICKET_STATES.FOR_APPROVAL}
            >
              {ticket.addIndividualTicketDetails.approveStatus
                ? t('Disapprove')
                : t('Approve')}
            </Button>
          )}
        </Box>
      </Title>
      <Grid container spacing={6}>
        {allLabels}
      </Grid>
    </ApproveBox>
  );
}
