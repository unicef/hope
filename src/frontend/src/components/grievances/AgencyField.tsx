import { Grid, IconButton } from '@mui/material';
import { Delete } from '@mui/icons-material';
import { useLocation } from 'react-router-dom';
import { Field } from 'formik';
import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { FormikSelectField } from '@shared/Formik/FormikSelectField';
import { FormikTextField } from '@shared/Formik/FormikTextField';
import { AllAddIndividualFieldsQuery } from '@generated/graphql';
import { getIndexForId } from './utils/helpers';

export interface AgencyFieldProps {
  id: string;
  baseName: string;
  baseNameArray?;
  onDelete;
  countryChoices: AllAddIndividualFieldsQuery['countriesChoices'];
  identityTypeChoices: AllAddIndividualFieldsQuery['identityTypeChoices'];
  isEdited?: boolean;
  values;
}

export function AgencyField({
  id,
  baseName,
  baseNameArray,
  onDelete,
  countryChoices,
  identityTypeChoices,
  isEdited,
  values,
}: AgencyFieldProps): React.ReactElement {
  const { t } = useTranslation();
  const agencyFieldName = `${baseName}.${getIndexForId(
    baseNameArray || values[baseName],
    id,
  )}`;

  const location = useLocation();
  const isEditTicket = location.pathname.indexOf('edit-ticket') !== -1;
  return (
    <Grid container alignItems="center" spacing={3}>
      <Grid item xs={4}>
        <Field
          name={`${agencyFieldName}.partner`}
          fullWidth
          variant="outlined"
          label={t('Partner')}
          component={FormikSelectField}
          choices={identityTypeChoices}
          required
          disabled={isEditTicket}
        />
      </Grid>
      <Grid item xs={4}>
        <Field
          name={`${agencyFieldName}.country`}
          fullWidth
          variant="outlined"
          label={t('Country')}
          component={FormikSelectField}
          choices={countryChoices}
          required
          disabled={isEditTicket}
        />
      </Grid>
      <Grid item xs={3}>
        <Field
          name={`${agencyFieldName}.number`}
          fullWidth
          variant="outlined"
          label={t('Identity Number')}
          component={FormikTextField}
          required
          disabled={isEditTicket}
        />
      </Grid>
      {!isEdited ? (
        <Grid item xs={1}>
          <IconButton disabled={isEditTicket} onClick={onDelete}>
            <Delete />
          </IconButton>
        </Grid>
      ) : null}
    </Grid>
  );
}
