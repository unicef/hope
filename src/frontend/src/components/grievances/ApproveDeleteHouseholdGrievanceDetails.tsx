import {
  Box,
  Button,
  DialogContent,
  DialogTitle,
  Grid,
  IconButton,
  Typography,
} from '@mui/material';
import styled from 'styled-components';
import Edit from '@mui/icons-material/Edit';
import { Field, Formik } from 'formik';
import * as React from 'react';
import { useState } from 'react';
import { useTranslation } from 'react-i18next';
import * as Yup from 'yup';
import { Dialog } from '@containers/dialogs/Dialog';
import { DialogActions } from '@containers/dialogs/DialogActions';
import { DialogContainer } from '@containers/dialogs/DialogContainer';
import { DialogFooter } from '@containers/dialogs/DialogFooter';
import { DialogTitleWrapper } from '@containers/dialogs/DialogTitleWrapper';
import { useSnackbar } from '@hooks/useSnackBar';
import { FormikRadioGroup } from '@shared/Formik/FormikRadioGroup';
import { FormikTextField } from '@shared/Formik/FormikTextField';
import { GRIEVANCE_TICKET_STATES } from '@utils/constants';
import {
  GrievanceTicketDocument,
  GrievanceTicketQuery,
  useApproveDeleteHouseholdDataChangeMutation,
} from '@generated/graphql';

const EditIcon = styled(Edit)`
  color: ${({ theme }) => theme.hctPalette.darkerBlue};
`;
export interface ApproveDeleteHouseholdGrievanceDetailsProps {
  ticket: GrievanceTicketQuery['grievanceTicket'];
  type: 'edit' | 'button';
}

export const ApproveDeleteHouseholdGrievanceDetails = ({
  ticket,
  type,
}: ApproveDeleteHouseholdGrievanceDetailsProps): React.ReactElement => {
  const { t } = useTranslation();
  const [dialogOpen, setDialogOpen] = useState(false);
  const [mutate] = useApproveDeleteHouseholdDataChangeMutation();
  const { showMessage } = useSnackbar();
  const isForApproval = ticket.status === GRIEVANCE_TICKET_STATES.FOR_APPROVAL;
  const { approveStatus, reasonHousehold } =
    ticket.deleteHouseholdTicketDetails;

  const validationSchema = Yup.object().shape({
    reasonHhId: Yup.string().when(
      'withdrawReason',
      (withdrawReasonValue: any) => {
        const value = String(withdrawReasonValue);
        if (value === 'duplicate' && !approveStatus) {
          return Yup.string()
            .required('Household Unicef Id is required')
            .max(15, 'Too long');
        }
        return Yup.string();
      },
    ),
  });

  const matchDialogTitle = (): string => {
    if (approveStatus && type !== 'edit') {
      return t('Warning');
    }
    return t('Reason of withdrawal');
  };

  const showWithdraw = (): boolean => {
    if (approveStatus && type === 'edit') {
      return true;
    }
    if (!approveStatus && type === 'button') {
      return true;
    }
    return false;
  };

  return (
    <Formik
      enableReinitialize
      initialValues={{
        withdrawReason: reasonHousehold ? 'duplicate' : 'other',
        reasonHhId: type === 'edit' ? reasonHousehold?.unicefId : '',
      }}
      validationSchema={validationSchema}
      onSubmit={async (values, { resetForm }) => {
        try {
          await mutate({
            variables: {
              grievanceTicketId: ticket.id,
              approveStatus: type === 'edit' ? true : !approveStatus,
              reasonHhId:
                values.withdrawReason === 'duplicate' ? values.reasonHhId : '',
            },
            refetchQueries: () => [
              {
                query: GrievanceTicketDocument,
                variables: { id: ticket.id },
              },
            ],
          });
          if (type === 'edit') {
            showMessage(t('Changes Approved'));
          }
          if (approveStatus) {
            showMessage(t('Changes Disapproved'));
          }
          if (!approveStatus) {
            showMessage(t('Changes Approved'));
          }
          setDialogOpen(false);
          resetForm();
        } catch (e) {
          e.graphQLErrors.map((x) => showMessage(x.message));
        }
      }}
    >
      {({ values, submitForm, resetForm }) => (
        <>
          <Box p={2}>
            {type === 'edit' ? (
              <IconButton
                data-cy="edit-button"
                onClick={() => setDialogOpen(true)}
              >
                <EditIcon />
              </IconButton>
            ) : (
              <Button
                data-cy="button-approve"
                onClick={() => setDialogOpen(true)}
                variant={approveStatus ? 'outlined' : 'contained'}
                color="primary"
                disabled={!isForApproval}
              >
                {approveStatus ? t('Disapprove') : t('Approve')}
              </Button>
            )}
          </Box>
          <Dialog
            open={dialogOpen}
            onClose={() => setDialogOpen(false)}
            scroll="paper"
            aria-labelledby="form-dialog-title"
            maxWidth="md"
          >
            <DialogTitleWrapper>
              <DialogTitle>{matchDialogTitle()}</DialogTitle>
            </DialogTitleWrapper>
            <DialogContent>
              <DialogContainer>
                <Box display="flex" flexDirection="column">
                  <Box mt={2}>
                    <Typography variant="body2">
                      {showWithdraw()
                        ? t(
                          'Please provide the reason of withdrawal of this household.',
                        )
                        : t(
                          'You did not approve the following household to be withdrawn. Are you sure you want to continue?',
                        )}
                    </Typography>
                  </Box>
                  {showWithdraw() && (
                    <Box>
                      <Field
                        name="withdrawReason"
                        choices={[
                          {
                            value: 'duplicate',
                            name: 'This household is a duplicate of another household',
                          },
                        ]}
                        component={FormikRadioGroup}
                        noMargin
                      />
                      {values.withdrawReason === 'duplicate' && (
                        <Grid container>
                          <Grid item xs={6}>
                            <Field
                              name="reasonHhId"
                              fullWidth
                              variant="outlined"
                              label={t('Household Unicef Id')}
                              component={FormikTextField}
                              required
                            />
                          </Grid>
                        </Grid>
                      )}
                      <Field
                        name="withdrawReason"
                        choices={[{ value: 'other', name: 'Other' }]}
                        component={FormikRadioGroup}
                        noMargin
                      />
                    </Box>
                  )}
                </Box>
              </DialogContainer>
            </DialogContent>
            <DialogFooter>
              <DialogActions>
                <Button
                  onClick={() => {
                    setDialogOpen(false);
                    resetForm();
                  }}
                >
                  {t('CANCEL')}
                </Button>
                <Button
                  type="submit"
                  color="primary"
                  variant="contained"
                  onClick={submitForm}
                  data-cy="button-submit"
                >
                  {t('Confirm')}
                </Button>
              </DialogActions>
            </DialogFooter>
          </Dialog>
        </>
      )}
    </Formik>
  );
};
