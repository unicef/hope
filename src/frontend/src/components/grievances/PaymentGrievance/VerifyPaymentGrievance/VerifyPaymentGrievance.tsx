import { Box, Button, DialogContent, DialogTitle, Grid } from '@mui/material';
import { Field, Form, Formik } from 'formik';
import * as React from 'react';
import { useState } from 'react';
import { useTranslation } from 'react-i18next';
import { Dialog } from '@containers/dialogs/Dialog';
import { DialogActions } from '@containers/dialogs/DialogActions';
import { DialogContainer } from '@containers/dialogs/DialogContainer';
import { DialogFooter } from '@containers/dialogs/DialogFooter';
import { DialogTitleWrapper } from '@containers/dialogs/DialogTitleWrapper';
import { useSnackbar } from '@hooks/useSnackBar';
import { FormikRadioGroup } from '@shared/Formik/FormikRadioGroup';
import { FormikTextField } from '@shared/Formik/FormikTextField';
import {
  GrievanceTicketDocument,
  GrievanceTicketQuery,
  useUpdateGrievanceMutation,
} from '@generated/graphql';
import { AutoSubmitFormOnEnter } from '@core/AutoSubmitFormOnEnter';

export interface VerifyPaymentGrievanceProps {
  ticket: GrievanceTicketQuery['grievanceTicket'];
}
export function VerifyPaymentGrievance({
  ticket,
}: VerifyPaymentGrievanceProps): React.ReactElement {
  const { t } = useTranslation();
  const [verifyManualDialogOpen, setVerifyManualDialogOpen] = useState(false);
  const { showMessage } = useSnackbar();
  const [mutate, { error }] = useUpdateGrievanceMutation();

  const submit = async (values): Promise<void> => {
    try {
      await mutate({
        variables: {
          input: {
            ticketId: ticket.id,
            extras: {
              ticketPaymentVerificationDetailsExtras: {
                newReceivedAmount: values.newReceivedAmount,
                newStatus: values.newStatus,
              },
            },
          },
        },
        refetchQueries: () => [
          {
            query: GrievanceTicketDocument,
            variables: { id: ticket.id },
          },
        ],
      });
    } catch (e) {
      e.graphQLErrors.map((x) => showMessage(x.message));
      return;
    }
    if (!error) {
      setVerifyManualDialogOpen(false);
      showMessage(t('Payment has been verified.'));
    }
  };

  const initialValues = {
    ticketId: ticket.id,
    newReceivedAmount: 0,
    newStatus: 'RECEIVED',
  };

  return (
    <Formik initialValues={initialValues} onSubmit={submit}>
      {({ values }) => (
        <Form>
          {verifyManualDialogOpen && <AutoSubmitFormOnEnter />}
          <Box p={2}>
            <Button
              color="primary"
              variant="contained"
              onClick={() => setVerifyManualDialogOpen(true)}
            >
              {t('Verify')}
            </Button>
          </Box>
          <Dialog
            open={verifyManualDialogOpen}
            onClose={() => setVerifyManualDialogOpen(false)}
            scroll="paper"
            aria-labelledby="form-dialog-title"
            maxWidth="md"
          >
            <DialogTitleWrapper>
              <DialogTitle>{t('Verify Payment')}</DialogTitle>
            </DialogTitleWrapper>
            <DialogContent>
              <DialogContainer>
                <Grid container>
                  <Grid item xs={12}>
                    <Field
                      name="newStatus"
                      label="Status"
                      style={{ flexDirection: 'row' }}
                      choices={[
                        { value: 'RECEIVED', name: t('Received') },
                        { value: 'NOT_RECEIVED', name: t('Not Received') },
                      ]}
                      component={FormikRadioGroup}
                    />
                  </Grid>
                  <Grid item xs={6}>
                    {values.newStatus === 'RECEIVED' && (
                      <Field
                        name="newReceivedAmount"
                        type="number"
                        label={t('Amount Received')}
                        color="primary"
                        component={FormikTextField}
                      />
                    )}
                  </Grid>
                </Grid>
              </DialogContainer>
            </DialogContent>
            <DialogFooter>
              <DialogActions>
                <Button onClick={() => setVerifyManualDialogOpen(false)}>
                  {t('CANCEL')}
                </Button>
                <Button
                  type="submit"
                  color="primary"
                  variant="contained"
                  onClick={() => submit(values)}
                  data-cy="button-submit"
                >
                  {t('Verify')}
                </Button>
              </DialogActions>
            </DialogFooter>
          </Dialog>
        </Form>
      )}
    </Formik>
  );
}
