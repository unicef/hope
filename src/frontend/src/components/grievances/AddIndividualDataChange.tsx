import { Button, Grid, Typography } from '@mui/material';
import { AddCircleOutline } from '@mui/icons-material';
import { useLocation } from 'react-router-dom';
import CalendarTodayRoundedIcon from '@mui/icons-material/CalendarTodayRounded';
import { Field, FieldArray } from 'formik';
import camelCase from 'lodash/camelCase';
import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { FormikDateField } from '@shared/Formik/FormikDateField';
import { FormikFileField } from '@shared/Formik/FormikFileField';
import { FormikSelectField } from '@shared/Formik/FormikSelectField';
import { FormikTextField } from '@shared/Formik/FormikTextField';
import {
  AllAddIndividualFieldsQuery,
  useAllAddIndividualFieldsQuery,
} from '@generated/graphql';
import { LoadingComponent } from '@core/LoadingComponent';
import { Title } from '@core/Title';
import { AgencyField } from './AgencyField';
import { DocumentField } from './DocumentField';
import { FormikBoolFieldGrievances } from './FormikBoolFieldGrievances';
import { removeItemById } from './utils/helpers';

export interface AddIndividualDataChangeFieldProps {
  field: AllAddIndividualFieldsQuery['allAddIndividualsFieldsAttributes'][number];
  flexField?: boolean;
}
export function AddIndividualDataChangeField({
  field,
  flexField,
}: AddIndividualDataChangeFieldProps): React.ReactElement {
  let fieldProps;
  const location = useLocation();
  const isEditTicket = location.pathname.indexOf('edit-ticket') !== -1;

  switch (field.type) {
    case 'DECIMAL':
      fieldProps = {
        component: FormikTextField,
      };
      break;
    case 'INTEGER':
      fieldProps = {
        component: FormikTextField,
        type: 'number',
      };
      break;
    case 'STRING':
      fieldProps = {
        component: FormikTextField,
      };
      break;
    case 'SELECT_ONE':
      fieldProps = {
        choices: field.choices,
        component: FormikSelectField,
      };
      break;
    case 'SELECT_MANY':
      fieldProps = {
        choices: field.choices,
        component: FormikSelectField,
        multiple: true,
      };
      break;
    case 'SELECT_MULTIPLE':
      fieldProps = {
        choices: field.choices,
        component: FormikSelectField,
      };
      break;
    case 'DATE':
      fieldProps = {
        component: FormikDateField,
        decoratorEnd: <CalendarTodayRoundedIcon color="disabled" />,
      };
      break;

    case 'BOOL':
      fieldProps = {
        component: FormikBoolFieldGrievances,
        required: field.required,
      };
      break;
    case 'IMAGE':
      fieldProps = {
        component: FormikFileField,
      };
      break;
    default:
      fieldProps = {};
  }
  return (
    <>
      <Grid item xs={8}>
        <Field
          name={`individualData${flexField ? '.flexFields' : ''}.${camelCase(
            field.name,
          )}`}
          fullWidth
          variant="outlined"
          label={field.labelEn}
          required={field.required}
          disabled={isEditTicket}
          {...fieldProps}
        />
      </Grid>
      <Grid item xs={4} />
    </>
  );
}

export interface AddIndividualDataChangeProps {
  values;
  setFieldValue?;
}

export function AddIndividualDataChange({
  values,
  setFieldValue,
}: AddIndividualDataChangeProps): React.ReactElement {
  const { t } = useTranslation();
  const location = useLocation();
  const isEditTicket = location.pathname.indexOf('edit-ticket') !== -1;

  const { data, loading } = useAllAddIndividualFieldsQuery();
  if (loading) {
    return <LoadingComponent />;
  }
  const flexFields = data.allAddIndividualsFieldsAttributes.filter(
    (item) => item.isFlexField,
  );
  const coreFields = data.allAddIndividualsFieldsAttributes.filter(
    (item) => !item.isFlexField,
  );

  return (
    !isEditTicket && (
      <>
        <Title>
          <Typography variant="h6">{t('Individual Data')}</Typography>
        </Title>
        <Grid container spacing={3}>
          <Grid item xs={12}>
            {t('Core Fields')}
          </Grid>

          {coreFields.map((item) => (
            <AddIndividualDataChangeField key={item.name} field={item} />
          ))}
          {flexFields.length > 0 && (
            <Grid item xs={12}>
              {t('Flex Fields')}
            </Grid>
          )}
          {flexFields.map((item) => (
            <AddIndividualDataChangeField
              key={item.name}
              field={item}
              flexField
            />
          ))}
        </Grid>
        <Grid container spacing={3}>
          <FieldArray
            name="individualData.documents"
            render={(arrayHelpers) => (
              <>
                {values.individualData?.documents?.map((item) => {
                  const existingOrNewId = item.node?.id || item.id;
                  return (
                    <DocumentField
                      id={existingOrNewId}
                      key={`${existingOrNewId}-${item?.country}-${item?.type?.key}`}
                      onDelete={() =>
                        removeItemById(
                          values.individualData.documents,
                          existingOrNewId,
                          arrayHelpers,
                        )
                      }
                      countryChoices={data.countriesChoices}
                      documentTypeChoices={data.documentTypeChoices}
                      baseName="individualData.documents"
                      baseNameArray={values.individualData.documents}
                      setFieldValue={setFieldValue}
                      values={values}
                    />
                  );
                })}
                <Grid item xs={8} />
                <Grid item xs={12}>
                  <Button
                    color="primary"
                    startIcon={<AddCircleOutline />}
                    disabled={isEditTicket}
                    onClick={() => {
                      arrayHelpers.push({
                        id: crypto.randomUUID(),
                        country: null,
                        key: null,
                        number: '',
                      });
                    }}
                  >
                    {t('Add Document')}
                  </Button>
                </Grid>
              </>
            )}
          />
        </Grid>
        <Grid container spacing={3}>
          <FieldArray
            name="individualData.identities"
            render={(arrayHelpers) => (
              <>
                {values.individualData?.identities?.map((item) => {
                  const existingOrNewId = item.node?.id || item.id;
                  return (
                    <AgencyField
                      key={existingOrNewId}
                      id={existingOrNewId}
                      onDelete={() =>
                        removeItemById(
                          values.individualData.identities,
                          existingOrNewId,
                          arrayHelpers,
                        )
                      }
                      countryChoices={data.countriesChoices}
                      identityTypeChoices={data.identityTypeChoices}
                      baseName="individualData.identities"
                      baseNameArray={values.individualData.identities}
                      values={values}
                    />
                  );
                })}
                <Grid item xs={8} />
                <Grid item xs={12}>
                  <Button
                    color="primary"
                    startIcon={<AddCircleOutline />}
                    onClick={() => {
                      arrayHelpers.push({
                        id: crypto.randomUUID(),
                        country: null,
                        partner: null,
                        number: '',
                      });
                    }}
                  >
                    {t('Add Identity')}
                  </Button>
                </Grid>
              </>
            )}
          />
        </Grid>
      </>
    )
  );
}
