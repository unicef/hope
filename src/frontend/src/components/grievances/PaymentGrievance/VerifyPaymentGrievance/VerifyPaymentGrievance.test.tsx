import * as React from 'react';
import { MockedProvider } from '@apollo/react-testing';
import { render } from '../../../../testUtils/testUtils';
import { fakeApolloUpdateGrievanceDetailsPaymentVerificationMutation } from '../../../../../fixtures/grievances/fakeApolloUpdateGrievanceDetailsPaymentVerificationMutation';
import { fakeGrievanceTicketPaymentVerification } from '../../../../../fixtures/grievances/fakeGrievanceTicketPaymentVerification';
import { VerifyPaymentGrievance } from './VerifyPaymentGrievance';

describe('components/grievances/VerifyPaymentGrievance', () => {
  it('should render with data', () => {
    const { container } = render(
      <MockedProvider
        addTypename={false}
        mocks={fakeApolloUpdateGrievanceDetailsPaymentVerificationMutation}
      >
        <VerifyPaymentGrievance
          ticket={fakeGrievanceTicketPaymentVerification}
        />
      </MockedProvider>,
    );
    expect(container).toMatchSnapshot();
  });
});
