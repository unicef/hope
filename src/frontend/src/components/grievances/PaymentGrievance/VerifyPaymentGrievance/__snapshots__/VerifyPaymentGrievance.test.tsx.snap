// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/grievances/VerifyPaymentGrievance should render with data 1`] = `
<div>
  <form
    action="#"
  >
    <div
      class="MuiBox-root css-hpgf8j"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-dl1jos-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Verify
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </form>
</div>
`;
